### DISCUSSION TITLE
Deprecating Cluster::Node$interface

### DISCUSSION THREAD

**Discussion by awelzel** [zeek/zeek#2877] [Ideas]
Not sure this is an idea, but not sure where else to put it.

Is the `Cluster::Node$interface` something that could be deprecated? In #2841 a bit of discussion happened that it isn't necessarily populated by `cluster-layout.zeek` generators today (zeekctl does it though) and things are generally working still as nothing relies on it.

Within base, [policy/misc/load-balancing](https://github.com/zeek/zeek/blob/master/scripts/policy/misc/load-balancing.zeek) does in fact use it, but today we may deprecate the script. Using BPF filters for load-balancing could as well be a package (and have a separate table of workers to interfaces).

@J-Gras pointed at [add-interfaces](https://github.com/J-Gras/add-interfaces) adding interfaces to logs, but that use-cases is probably better handled with the `packet_source()` bif today (and hasn't worked in non-zeekctl environments).

Do people think putting a `&deprecated` on it willcause uproar? Happy to cross-post on community.zeek.org.

**Comment by ckreibich** [comment]
It does strike me as useful for the script layer to know the interface a node/process is working with. Are you asking because you think it's not useful, or because it's already available elsewhere?

I have pondered this in the past from the other end, because it's also mentioned in the Supervisor framework [here](https://github.com/zeek/zeek/blob/master/scripts/base/frameworks/supervisor/api.zeek#L27-L29) and [here](https://github.com/zeek/zeek/blob/master/scripts/base/frameworks/supervisor/api.zeek#L37-L38).

**Reply by awelzel** [reply]
> It does strike me as useful for the script layer to know the interface a node/process is working with. Are you asking because you think it's not useful, or because it's already available elsewhere?

For the local process, that can be done via the [packet_source](https://docs.zeek.org/en/master/scripts/base/bif/zeek.bif.zeek.html#id-packet_source) bif today (which is more powerful and even gives you info when there's no `$interface` in cluster-layout or `$interface` has been to something inconsistent, or you're not running in cluster mode anyway).

If you want a global view, cluster-layout can be convenient, but could also be argued that it can then be injected/templated by users manually.

> I have pondered this in the past from the other end, because it's also mentioned in the Supervisor framework [here]

Yeah, the supervisor needs it in the `NodeConfig` to give a process an interface to listen on, that we can't let out. It could be removed from the `ClusterEndpoint` because that's used to populate the `Cluster::nodes` field. The [supervisor docs](https://docs.zeek.org/en/master/frameworks/supervisor.html#supervised-cluster-example) structure this a bit differently and it looks like the interface on ClusterEndpoint is needed, but I believe it's not technically.

It's mainly that `Node$interface` is not guaranteed to be set and not required for most deployments.

**Reply by ckreibich** [reply]
Yeah, that all makes sense to me, so I'm in favor. If there are use cases for node A needing to know the interface on node B, that would be interesting to hear.

> It could be removed from the `ClusterEndpoint`

Yep. This would simplify some Management Framework code, too.

Btw there's another emerging deprecation in `Cluster::Node`, for transitioning the `ip` field to something also supporting naming (#2729).
